File: packages/engine/src/travel.ts
/**
 * ⭐ MESAFE ve SEFER SÜRESİ (SİSTEM PLANI §13.5)
 *
 * Referans uygulama `harita.html`; buradaki formüller onunla BİREBİR aynıdır (§13.5.5 cetveli
 * bu koddan da çıkar). Motor gibi bu modül de SAF: DB, zaman, IO bilmez.
 *
 *   D = Δşehir + U·Δdiyar + W·Δkıta                 U = 20, W = 4000  (Manhattan/toplamalı)
 *   T = (TABAN + geçiş + K · D^p / (1 + 0,05·Haritacılık)) · (100/v)      tavan 24 saat
 *
 * ⭐⭐ **HER ŞEY HIZLA ORANTILI** (kullanıcı kararı, 2026-08-03). Taban da `100/v` ile bölünür ve
 * casus kuşun **ayrı tabanı yoktur** — kuşun 6000 hızı farkı kendisi yaratır. Eski hâlde taban
 * hızdan bağımsızdı (ordu 600 sn / casus 120 sn) ve sonuç şuydu: 60 kat hızlı olan kuş komşu
 * şehre 2 dk 10 sn'de, ordu 20 dk'da gidiyordu — yani **hız birimi yok sayılıyordu**. Artık
 * kuşun süresi ordununkinin tam 1/60'ı.
 *
 * ⚠️ **Haritacılık YALNIZ yol terimini kısaltır** (taban ve geçiş ek süresi hariç, §13.5.3).
 * Baskın–savunma dengesinin ayar vidası budur: her şeyi bölseydi Haritacılık 15'te komşu
 * baskını 23 dakikaya inerdi ve erken oyundaki "hızlı yağma → hızlı gelişme" sarmalı geri
 * gelirdi (kullanıcı bu seçeneği bilerek almadı).
 *
 * İstemci aynı fonksiyonu YALNIZ önizleme için kullanır; otorite `execute_at` yazan sunucudur.
 */
import { HERO_SPEED, SPEED_EXEMPT_WHEN_ESCORTED, UNITS_BY_ID } from '@mobilwar/catalog';

export interface Coordinates {
  /** kıta */
  k: number;
  /** diyar */
  d: number;
  /** şehir yeri */
  s: number;
}

export interface MapConfig {
  /** Bir diyar farkının kaç "şehir" ettiği. */
  districtWeight: number;
  /** Bir kıta farkının kaç "şehir" ettiği (1 kıta = 200 diyar). */
  continentWeight: number;
  /** Yol terimi katsayısı. */
  k: number;
  /** Mesafe sıkıştırma üssü: mesafe 100× artınca süre ~7× artar. */
  p: number;
  /**
   * Taban süre (sn) — "orduyu toplayıp yola çıkarmak", hız 100 için.
   * ⚠️ Tek taban: casus kuşun ayrı tabanı YOK, hızı farkı kendisi yaratıyor.
   */
  baseSeconds: number;
  /** Diyar değiştiren sefere eklenen süre (sn, hız 100 için). Haritacılık kısaltmaz. */
  districtCrossSeconds: number;
  /** Kıta değiştiren sefere eklenen süre (sn, hız 100 için). Haritacılık kısaltmaz. */
  continentCrossSeconds: number;
  /** Süre tavanı (sa) — zıt köşe bile bunu aşmaz. */
  capHours: number;
  /** Haritacılık seviye başına hız kazancı. */
  cartographyStep: number;
  /**
   * ⭐⭐ YÜK ARABASI KAFİLEYİ YAVAŞLATMASIN (kullanıcı, 2026-08-21) — dünya ayarı.
   *
   * Açıkken (**varsayılan**) `SPEED_EXEMPT_WHEN_ESCORTED` birimleri, orduda başka yürüyen
   * birim varken en-yavaş hesabına girmiyor: araba kafilenin hızına ayak uyduruyor.
   * Kapalıyken davranış 2026-08-21 öncesiyle **bit-bit aynı**: arabanın 140 hızı da hesaba
   * giriyor ve bir ejderha ordusunu (160) 140'a çekiyor.
   *
   * ⚠️ Muafiyet **refakat şartlı**: araba yalnız başına gönderildiğinde (nakliye · destek ·
   * şehir kurma) yine kendi hızıyla yürüyor — yoksa "hızı olmayan birim" üretirdik.
   * ⚠️ Kahraman refakatçi SAYILMAZ: tek arabayla giden bir kahraman, arabayı kendi hızına
   * (200) çekseydi araba fiziksel olarak imkânsız bir hızda ilerlemiş olurdu.
   */
  cargoIgnoresSpeed: boolean;
}

/**
 * ⚠️ Sayılar 2026-08-03'te **iki katına** çıktı: komşu şehre sefer 20 dk'ydı ve kullanıcı bunu
 * erken oyunda *"hızlı yağma → hızlı gelişme"* sarmalı olarak gördü. Hedef 35-45 dk, seçilen
 * **40 dk**. Komşu şehirde `D=1` ve `1^p = 1` olduğu için mesafe terimi hiç iş yapmaz —
 * 40 dakikayı belirleyen tek şey `baseSeconds + k`. `p` yalnız uzağı şekillendirir ve
 * 0,46 → 0,42'ye indi ki K iki katına çıkınca kıtalar arası süreler tavana yapışmasın.
 */
export const DEFAULT_MAP_CONFIG: MapConfig = {
  districtWeight: 20,
  continentWeight: 4000,
  k: 1200,
  p: 0.42,
  baseSeconds: 1200,
  // ⚠️ Varsayılan 0 — bugün geçişin ek bir bedeli YOK. Kullanıcı bunu ileride açmak için
  //    ayarlanabilir istedi; sıfır bırakıldığı sürece cetvel yalnız `baseSeconds + k`'dan çıkar.
  districtCrossSeconds: 0,
  continentCrossSeconds: 0,
  capHours: 24,
  cartographyStep: 0.05,
  // ⭐ Varsayılan AÇIK (kullanıcı kararı, 2026-08-21).
  cargoIgnoresSpeed: true,
};

/** Kısmi override'ı varsayılanla birleştirir (`mergeCombatConfig` deseni). */
export function mergeMapConfig(patch?: Partial<MapConfig>): MapConfig {
  return patch ? { ...DEFAULT_MAP_CONFIG, ...patch } : DEFAULT_MAP_CONFIG;
}

/**
 * Kademeli/toplamalı mesafe. **Öklid DEĞİL**: her koordinat basamağındaki fark süreye mutlaka
 * yansır, yoksa "1 kıta + 200 diyar" ile "1 kıta" aynı süreye inerdi (§13.5.1).
 */
export function distance(a: Coordinates, b: Coordinates, cfg: MapConfig = DEFAULT_MAP_CONFIG): number {
  return Math.abs(a.s - b.s)
    + cfg.districtWeight * Math.abs(a.d - b.d)
    + cfg.continentWeight * Math.abs(a.k - b.k);
}

/**
 * Mesafe + geçiş bayrakları tek çağrıda.
 *
 * ⚠️ Bayrakları her çağıranın kendisi (`a.d !== b.d`) hesaplamasındansa burada üretiliyor:
 * dört ayrı çağrı noktası var (saldırı · yürüyüş · kahraman dönüşü · istemci önizlemesi) ve
 * birinde unutulursa **önizleme ile gerçek varış anı sessizce ayrışır**.
 */
export function route(a: Coordinates, b: Coordinates, cfg: MapConfig = DEFAULT_MAP_CONFIG): {
  distance: number; crossesDistrict: boolean; crossesContinent: boolean;
} {
  return {
    distance: distance(a, b, cfg),
    crossesDistrict: a.d !== b.d,
    crossesContinent: a.k !== b.k,
  };
}

/**
 * Ordunun hızı = **en yavaş üyenin** hızı.
 *
 * ⭐ **KAHRAMAN DA ÜYEDİR** (kullanıcı, 2026-08-03). Kural bir süre *"kahraman orduyu
 * hızlandırmaz, bu yüzden hesaba hiç girmez"* diye yazılıydı ve pratikte doğru sonuç
 * veriyordu — çünkü kahramanın hızı (200) her savaşçıdan yüksekti (80-160), yani en yavaş
 * hiçbir zaman o olmuyordu.
 *
 * ⚠️ Casus Kuş **6000** hızıyla bu varsayımı kırıyor. Kuş destek görevine katılabilir hâle
 * gelince (aynı gün) "9 kuş + 1 kahraman" ordusu ortaya çıktı ve önizleme **52 saniye**
 * gösterdi: kahraman ışınlanmış gibi. Doğrusu 200, yani kuş kahramanı bekler.
 * Yani kural değişmedi — *"kahraman hızlandırmaz"* hâlâ geçerli; eksik olan **yavaşlatır**
 * tarafıydı ve o zamana kadar hiç gözlemlenememişti.
 *
 * Bilinmeyen birim id'si veya yürüyemeyen (hız 0) birim varsa `null` döner — çağıran bunu
 * doğrulama hatası olarak işler.
 */
export function armySpeed(
  counts: Record<string, number>, heroCount = 0, cfg: MapConfig = DEFAULT_MAP_CONFIG,
): number | null {
  /**
   * ⭐⭐ İKİ ADAY BİRDEN (2026-08-21): tüm birimlerin en yavaşı ve **muaflar hariç** en yavaşı.
   *
   * ⚠️ Tek geçişte ikisini de toplamak şart, çünkü hangisinin kullanılacağı ancak döngü
   * bittiğinde belli oluyor: muaf olmayan hiç birim yoksa (yalnız yük arabası gönderilmiş)
   * muafiyet **düşüyor** ve arabanın kendi hızı geçerli oluyor.
   * ⚠️ Geçersizlik kapısı (`speed <= 0` → `null`) muaf birimler için de işliyor: muafiyet
   * "hızı yok sayılır" demek, "birim geçerlidir" demek değil.
   */
  let slowestAll = Infinity;
  let slowestEscort = Infinity;
  for (const [id, n] of Object.entries(counts)) {
    if (!(n > 0)) continue;
    const speed = UNITS_BY_ID[id]?.speed ?? 0;
    if (speed <= 0) return null;
    if (speed < slowestAll) slowestAll = speed;
    if (SPEED_EXEMPT_WHEN_ESCORTED.has(id)) continue;
    if (speed < slowestEscort) slowestEscort = speed;
  }

  /* Muafiyet YALNIZ refakatçi varken: `slowestEscort` sonsuzsa orduda muaf olmayan birim yok. */
  let slowest = cfg.cargoIgnoresSpeed && Number.isFinite(slowestEscort)
    ? slowestEscort
    : slowestAll;

  if (heroCount > 0 && HERO_SPEED < slowest) slowest = HERO_SPEED;
  return Number.isFinite(slowest) ? slowest : null;
}

export interface TravelInput {
  /** `distance()` çıktısı. */
  distance: number;
  /** Ordunun hızı (`armySpeed()`); casus seferinde casus kuşun hızı. */
  speed: number;
  /** Saldıranın Haritacılık seviyesi. */
  cartography?: number;
  /** Sefer diyar değiştiriyor mu (`districtCrossSeconds` eklenir)? */
  crossesDistrict?: boolean;
  /** Sefer kıta değiştiriyor mu (`continentCrossSeconds` eklenir)? */
  crossesContinent?: boolean;
  /** Dünya hız çarpanı — TÜM süreyi böler (`worlds.speed_multiplier`). */
  speedMultiplier?: number;
}

/**
 * Sefer süresi (saniye, yukarı yuvarlanmış tam sayı).
 *
 * Dönüş bacağı da AYNI süredir; görev tipi süreyi değiştirmez (saldırı = destek = nakliye = casus).
 *
 * ⚠️ Görev tipine bakan tek bir dal bile YOK — casus seferi de bu fonksiyondan geçer ve
 * farkını yalnız Casus Kuş'un hızından (6000) alır. Buraya tipe özel bir sabit eklemek,
 * katalogdaki hız sütununu yeniden anlamsızlaştırır.
 */
export function travelSeconds(input: TravelInput, cfg: MapConfig = DEFAULT_MAP_CONFIG): number {
  const speed = Math.max(1, input.speed);
  const cartographyFactor = 1 + cfg.cartographyStep * Math.max(0, input.cartography ?? 0);
  const road = cfg.k * Math.max(0, input.distance) ** cfg.p / cartographyFactor;
  const cross = (input.crossesDistrict ? cfg.districtCrossSeconds : 0)
    + (input.crossesContinent ? cfg.continentCrossSeconds : 0);
  // ⭐ Taban ve geçiş de `100/v` ile ölçeklenir → süre baştan sona hıza orantılı.
  const total = (cfg.baseSeconds + cross + road) * (100 / speed);
  const capped = Math.min(total, cfg.capHours * 3600);
  // Hızlı dünya seçeneği süreyi böler; tavan da aynı oranda iner (§13.5.6).
  return Math.max(1, Math.ceil(capped / Math.max(0.01, input.speedMultiplier ?? 1)));
}

File: apps/mobile/lib/core/travel.dart
/// ⭐⭐ MESAFE ve SEFER SÜRESİ — `packages/engine/src/travel.ts`in Dart karşılığı (§13.5).
///
///   D = Δşehir + U·Δdiyar + W·Δkıta                 U = 20, W = 4000  (toplamalı)
///   T = (TABAN + geçiş + K · D^p / (1 + 0,05·Haritacılık)) · (100/v)      tavan 24 saat
///
/// ⚠️⚠️ **Davranış vektörle kilitli**: `packages/contracts/fixtures/travel-vectors.json` hem
/// TS hem Dart testinden okunuyor. Saat portunda (`clock.dart`) olduğu gibi, «tam eşitlik»
/// umut değil kapı — üstelik burada kaçış daha sinsi olurdu: `pow` iki dilde son basamakta
/// ayrışabilir ve ekrandaki süre sunucununkinden bir saniye sapabilir.
///
/// ⚠️ İstemci bu hesabı YALNIZ önizleme için yapar; otorite `execute_at` yazan sunucudur.
///
/// ─ ⛔ WEB'DEN AYRILAN TEK YER: HIZ NEREDEN GELİYOR ───────────────────────────────────────
/// Web'in `armySpeed`i birim hızlarını `UNITS_BY_ID`den, yani derlenmiş katalogdan okuyor.
/// **Dart'ta bu yapılamaz**: katalog Dart'a üretilmiyor çünkü değerleri dünya başına
/// override edilebiliyor (`catalog_model.dart` başlığı). Bu yüzden hızlar dışarıdan, sunucudan
/// gelen katalogdan geçiriliyor. Aynı kural, aynı sonuç, farklı kaynak — ve bu fark bilerek.
library;

import 'dart:math' as math;

/// ⭐ Kahramanın hızı. ⚠️ Sunucuda da düz bir sabit (`packages/catalog` · `HERO_SPEED`) ve
/// dünya ayarlarından geçmiyor — `travel.ts` onu doğrudan import ediyor. Ayarlanabilir hâle
/// gelirse buraya değil, **sunucu yanıtına** taşınmalı.
const int kHeroSpeed = 200;

/// Harita sabitleri.
///
/// ⚠️⚠️ Varsayılanlar burada duruyor ama **istemci onlara güvenmemeli**: panelden dünya başına
/// ayarlanabiliyorlar ve şehir yanıtı (`/cities/:id` → `map`) gerçek değerleri gönderiyor.
/// Varsayılan yalnız alan hiç gelmediğinde devreye giriyor.
class MwMapConfig {
  const MwMapConfig({
    this.districtWeight = 20,
    this.continentWeight = 4000,
    this.k = 1200,
    this.p = 0.42,
    this.baseSeconds = 1200,
    this.districtCrossSeconds = 0,
    this.continentCrossSeconds = 0,
    this.capHours = 24,
    this.cartographyStep = 0.05,
  });

  /// Bir diyar farkının kaç «şehir» ettiği.
  final num districtWeight;

  /// Bir kıta farkının kaç «şehir» ettiği.
  final num continentWeight;

  /// Yol terimi katsayısı.
  final num k;

  /// Mesafe sıkıştırma üssü.
  final num p;

  /// Taban süre (sn), hız 100 için. ⚠️ Casus kuşun AYRI tabanı yok; farkı hızı yaratıyor.
  final num baseSeconds;

  /// Diyar/kıta değiştiren sefere eklenen süre. ⚠️ Haritacılık bunları KISALTMAZ.
  final num districtCrossSeconds;
  final num continentCrossSeconds;

  /// Süre tavanı (saat).
  final num capHours;

  /// Haritacılık seviye başına hız kazancı.
  final num cartographyStep;

  static const MwMapConfig defaults = MwMapConfig();

  /// Sunucudan gelen kısmi override'ı varsayılanla birleştirir (`mergeMapConfig` deseni).
  ///
  /// ⚠️ Alan alan okunuyor, toptan değil: sunucu yeni bir alan eklerse eski istemci onu
  /// görmezden gelir ve **çökmez**; eksik bırakırsa varsayılana düşer.
  static MwMapConfig fromJson(Object? raw) {
    if (raw is! Map) return defaults;
    num at(String key, num fallback) {
      final v = raw[key];
      return v is num ? v : fallback;
    }

    const d = defaults;
    return MwMapConfig(
      districtWeight: at('districtWeight', d.districtWeight),
      continentWeight: at('continentWeight', d.continentWeight),
      k: at('k', d.k),
      p: at('p', d.p),
      baseSeconds: at('baseSeconds', d.baseSeconds),
      districtCrossSeconds: at('districtCrossSeconds', d.districtCrossSeconds),
      continentCrossSeconds: at(
        'continentCrossSeconds',
        d.continentCrossSeconds,
      ),
      capHours: at('capHours', d.capHours),
      cartographyStep: at('cartographyStep', d.cartographyStep),
    );
  }
}

typedef MwCoordinates = ({int k, int d, int s});

/// Kademeli/toplamalı mesafe. **Öklid DEĞİL**: her koordinat basamağındaki fark süreye mutlaka
/// yansımalı, yoksa «1 kıta + 200 diyar» ile «1 kıta» aynı süreye inerdi (§13.5.1).
num distance(
  MwCoordinates a,
  MwCoordinates b, [
  MwMapConfig cfg = MwMapConfig.defaults,
]) =>
    (a.s - b.s).abs() +
    cfg.districtWeight * (a.d - b.d).abs() +
    cfg.continentWeight * (a.k - b.k).abs();

typedef MwRoute = ({num distance, bool crossesDistrict, bool crossesContinent});

/// Mesafe + geçiş bayrakları tek çağrıda.
///
/// ⚠️ Bayrakları çağıranın kendisi (`a.d != b.d`) hesaplamıyor: sunucuda dört ayrı çağrı
/// noktası var ve birinde unutulursa **önizleme ile gerçek varış anı sessizce ayrışır**.
MwRoute route(
  MwCoordinates a,
  MwCoordinates b, [
  MwMapConfig cfg = MwMapConfig.defaults,
]) => (
  distance: distance(a, b, cfg),
  crossesDistrict: a.d != b.d,
  crossesContinent: a.k != b.k,
);

/// Ordunun hızı = **en yavaş üyenin** hızı. Hesaplanamıyorsa `null`.
///
/// ⭐⭐ **KAHRAMAN DA ÜYEDİR.** Kural bir süre *"kahraman orduyu hızlandırmaz, hesaba hiç
/// girmez"* diye yazılıydı ve pratikte doğru sonuç veriyordu — kahramanın 200'ü her savaşçıdan
/// (80-160) hızlıydı, yani en yavaş hiç o olmuyordu. **Casus Kuş'un 6000'i varsayımı kırdı**:
/// «9 kuş + 1 kahraman» ordusu önizlemede 52 saniyede varıyor göründü. Doğrusu 200 — kuş
/// kahramanı bekler. Kural değişmedi, eksik olan **yavaşlatır** tarafıydı.
///
/// ⚠️ Hız çözümü **dışarıdan** (`speedOf`): katalog Dart'a üretilmiyor, gerekçe dosya başlığında.
/// ⚠️ Bilinmeyen id ya da yürüyemeyen birim (hız ≤ 0) → `null`. Çağıran bunu doğrulama hatası
/// sayar; sessizce 0 kabul etmek sonsuz süre üretirdi.
int? armySpeed(
  Map<String, int> counts,
  int? Function(String id) speedOf, {
  int heroCount = 0,
}) {
  int? enYavas;
  for (final e in counts.entries) {
    if (e.value <= 0) continue;
    final s = speedOf(e.key) ?? 0;
    if (s <= 0) return null;
    if (enYavas == null || s < enYavas) enYavas = s;
  }
  if (heroCount > 0 && (enYavas == null || kHeroSpeed < enYavas)) {
    enYavas = kHeroSpeed;
  }
  return enYavas;
}

/// Sefer süresi (saniye, yukarı yuvarlanmış).
///
/// Dönüş bacağı da AYNI süredir; **görev tipi süreyi değiştirmez** (saldırı = destek = nakliye
/// = casus). ⚠️ Tipe bakan tek bir dal bile yok: casus seferi de buradan geçer ve farkını
/// yalnız Casus Kuş'un hızından alır. Tipe özel bir sabit eklemek, katalogdaki hız sütununu
/// yeniden anlamsızlaştırırdı.
///
/// ⚠️ **Haritacılık YALNIZ yol terimini kısaltır** (taban ve geçiş hariç, §13.5.3). Baskın-
/// savunma dengesinin ayar vidası budur; her şeyi bölseydi Haritacılık 15'te komşu baskını
/// 23 dakikaya inerdi.
int travelSeconds({
  required num distance,
  required num speed,
  num cartography = 0,
  bool crossesDistrict = false,
  bool crossesContinent = false,
  num speedMultiplier = 1,
  MwMapConfig cfg = MwMapConfig.defaults,
}) {
  final v = math.max(1, speed);
  final haritacilik = 1 + cfg.cartographyStep * math.max(0, cartography);
  final yol = cfg.k * math.pow(math.max(0, distance), cfg.p) / haritacilik;
  final gecis =
      (crossesDistrict ? cfg.districtCrossSeconds : 0) +
      (crossesContinent ? cfg.continentCrossSeconds : 0);
  // ⭐ Taban ve geçiş de `100/v` ile ölçekleniyor → süre baştan sona hıza orantılı.
  final toplam = (cfg.baseSeconds + gecis + yol) * (100 / v);
  final kelepce = math.min(toplam, cfg.capHours * 3600);
  // Hızlı dünya seçeneği süreyi böler; tavan da aynı oranda iner (§13.5.6).
  return math.max(1, (kelepce / math.max(0.01, speedMultiplier)).ceil());
}
